visdialch/data/readers.py

```diff
@@ -34,97 +35,183 @@ class DialogsReader(object):
+            self.answers = {
+                i: answer for i, answer in
+                enumerate(visdial_data["data"]["answers"] + [""])
+            }
 
-            # Add empty question, answer at the end, useful for padding dialog
-            # rounds for test.
-            self.questions.append("")
-            self.answers.append("")
+            # ``image_id``` serves as key for all three dicts here.
+            self.captions: Dict[int, Any] = {}
+            self.dialogs: Dict[int, Any] = {}
+            self.num_rounds: Dict[int, Any] = {}
 
-            # Image_id serves as key for all three dicts here.
-            self.captions = {}
-            self.dialogs = {}
-            self.num_rounds = {}
+            all_dialogs = visdial_data["data"]["dialogs"]
 
-            for dialog_for_image in visdial_data["data"]["dialogs"]:
-                self.captions[dialog_for_image["image_id"]] = dialog_for_image[
-                    "caption"
-                ]
+            # Retain only first ``num_examples`` dialogs if specified.
+            if num_examples is not None:
+                all_dialogs = all_dialogs[:num_examples]
+
+            for _dialog in all_dialogs:
+
+                self.captions[_dialog["image_id"]] = _dialog["caption"]
 
                 # Record original length of dialog, before padding.
                 # 10 for train and val splits, 10 or less for test split.
-                self.num_rounds[dialog_for_image["image_id"]] = len(
-                    dialog_for_image["dialog"]
-                )
+                self.num_rounds[_dialog["image_id"]] = len(_dialog["dialog"])
 
                 # Pad dialog at the end with empty question and answer pairs
                 # (for test split).
-                while len(dialog_for_image["dialog"]) < 10:
-                    dialog_for_image["dialog"].append(
-                        {"question": -1, "answer": -1}
-                    )
+                while len(_dialog["dialog"]) < 10:
+                    _dialog["dialog"].append({"question": -1, "answer": -1})
 
-                # Add empty answer /answer options if not provided
+                # Add empty answer (and answer options) if not provided
                 # (for test split).
-                for i in range(len(dialog_for_image["dialog"])):
-                    if "answer" not in dialog_for_image["dialog"][i]:
-                        dialog_for_image["dialog"][i]["answer"] = -1
-                    if "answer_options" not in dialog_for_image["dialog"][i]:
-                        dialog_for_image["dialog"][i]["answer_options"] = [
-                            -1
-                        ] * 100
-
-                self.dialogs[dialog_for_image["image_id"]] = dialog_for_image[
-                    "dialog"
-                ]
-
+                for i in range(len(_dialog["dialog"])):
+                    if "answer" not in _dialog["dialog"][i]:
+                        _dialog["dialog"][i]["answer"] = -1
```

Comment:
maybe add a comment here that this `-1` is being used to index into the answer dicts

Reply:
This will actually break on test split. Thanks for pointing it out!